internal/cmd/agent/register.go

```diff
@@ -0,0 +1,87 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	command "github.com/rancher/fleet/internal/cmd"
+	"github.com/rancher/fleet/internal/cmd/agent/register"
+	"github.com/rancher/wrangler/v2/pkg/kubeconfig"
+	"github.com/rancher/wrangler/v2/pkg/leader"
+	"github.com/rancher/wrangler/v2/pkg/ratelimit"
+)
+
+func NewRegister() *cobra.Command {
+	cmd := command.Command(&Register{}, cobra.Command{
+		Use:   "register [flags]",
+		Short: "Register agent with an upstream cluster",
+	})
+	return cmd
+}
+
+type Register struct {
+	command.DebugConfig
+	UpstreamOptions
+}
+
+func (r *Register) Run(cmd *cobra.Command, args []string) error {
+	// provide a logger in the context to be compatible with controller-runtime
+	zopts := zap.Options{
+		Development: true,
+	}
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&zopts)))
+	ctx := log.IntoContext(cmd.Context(), ctrl.Log)
+
+	if err := r.SetupDebug(); err != nil {
+		return fmt.Errorf("failed to setup debug logging: %w", err)
+	}
+
+	clientConfig := kubeconfig.GetNonInteractiveClientConfig(r.Kubeconfig)
+	kc, err := clientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	// try to claim leadership lease without rate limiting
+	localConfig := rest.CopyConfig(kc)
+	localConfig.RateLimiter = ratelimit.None
+	k8s, err := kubernetes.NewForConfig(localConfig)
+	if err != nil {
+		return err
+	}
+
+	setupLog.Info("starting registration on upstream cluster", "namespace", r.Namespace)
+
+	ctx, cancel := context.WithCancel(ctx)
+	leader.RunOrDie(ctx, r.Namespace, "fleet-agent-register-lock", k8s, func(ctx context.Context) {
```

Comment:
Does `leaderelection` make sense anymore?

Reply:
I wonder if we get anything from it. The code runs in an init container. 
Do we agree to keep the statefulset? Then I could remove all leader election from the agent, I think.